File: src/main/java/me/itzisonn_/meazy/parser/pasing_function/statement/ProgramParsingFunction.java
package me.itzisonn_.meazy.parser.pasing_function.statement;

import me.itzisonn_.meazy.MeazyMain;
import me.itzisonn_.meazy.parser.ast.program.impl.ProgramImpl;
import me.itzisonn_.meazy.parser.ParsingContext;
import me.itzisonn_.meazy.text.TextKt;
import me.itzisonn_.meazy.lexer.TokenTypes;
import me.itzisonn_.meazy.parser.Parser;
import me.itzisonn_.meazy.parser.UnexpectedTokenException;
import me.itzisonn_.meazy.parser.ast.program.Program;
import me.itzisonn_.meazy.parser.ast.statement.Statement;
import me.itzisonn_.meazy.parser.pasing_function.AbstractParsingFunction;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

@NullMarked
public class ProgramParsingFunction extends AbstractParsingFunction<Program> {
    public ProgramParsingFunction() {
        super("program");
    }

    @Override
    public Program parse(ParsingContext context, @Nullable Object... extra) {
        File file;
        if (extra.length == 0) throw new IllegalArgumentException("Expected file as extra argument");
        if (extra[0] == null) file = null;
        else if (extra[0] instanceof File fileArg) file = fileArg;
        else throw new IllegalArgumentException("Expected file as extra argument");

        Parser parser = context.getParser();
        parser.moveOverOptionalNewLines();

        List<Statement> body = new ArrayList<>();

        Statement headerStatement;

        while (true) {
            try {
                headerStatement = parser.parse(MeazyMain.getDefaultIdentifier("header_statement"), Statement.class);
            }
            catch (UnexpectedTokenException e) {
                break;
            }

            parser.next(TokenTypes.NEW_LINE(),  TextKt.translatable("meazy:parser.expected", "new_line"));
            parser.moveOverOptionalNewLines();

            body.add(headerStatement);
        }

        parser.moveOverOptionalNewLines();

        while (!parser.getCurrent().getType().equals(TokenTypes.END_OF_FILE())) {
            body.add(parser.parse(MeazyMain.getDefaultIdentifier("global_statement"), Statement.class));

            if (!parser.getCurrent().getType().equals(TokenTypes.END_OF_FILE())) {
                parser.next(TokenTypes.NEW_LINE(), TextKt.translatable("meazy:parser.expected", "new_line"));
                parser.moveOverOptionalNewLines();
            }
        }

        return new ProgramImpl(file, MeazyMain.INSTANCE.getVERSION(), body);
    }
}

File: src/main/java/me/itzisonn_/meazy/parser/pasing_function/statement/ProgramParsingFunction.kt
package me.itzisonn_.meazy.parser.pasing_function.statement

import me.itzisonn_.meazy.MeazyMain.VERSION
import me.itzisonn_.meazy.lexer.TokenTypes.endOfFile
import me.itzisonn_.meazy.lexer.TokenTypes.newLine
import me.itzisonn_.meazy.parser.Parser
import me.itzisonn_.meazy.parser.UnexpectedTokenException
import me.itzisonn_.meazy.parser.ast.statement.Program
import me.itzisonn_.meazy.parser.ast.statement.Statement
import me.itzisonn_.meazy.parser.pasing_function.ParsingFunction
import me.itzisonn_.meazy.text.translatable
import java.io.File

object ProgramParsingFunction : ParsingFunction<Program>("program") {
    override fun Parser.parse(vararg extra: Any?): Program {
        val file: File?
        require(extra.isNotEmpty()) { "Expected file as extra argument" }
        if (extra[0] is File) file = extra[0] as File
        else throw IllegalArgumentException("Expected file as extra argument")

        skipNewLines()

        val body = mutableListOf<Statement>()
        var headerStatement: Statement?

        while (true) {
            try {
                headerStatement = parse(HeaderStatementParsingFunction)
            }
            catch (_: UnexpectedTokenException) {
                break
            }

            next(newLine, translatable("meazy:parser.expected", "new_line"))
            skipNewLines()

            body.add(headerStatement)
        }

        skipNewLines()

        while (current.type != endOfFile) {
            body.add(parse(GlobalStatementParsingFunction))

            if (current.type != endOfFile) {
                next(newLine, translatable("meazy:parser.expected", "new_line"))
                skipNewLines()
            }
        }

        return Program(file, VERSION, body)
    }
}
